[PT2][Quant] Move to module partitioner for linear pattern quantization

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101122
* #101121
* #101120

Subgraph matcher is somewhat unreliable as the pattern can vary depending on
the dimensionality of input tensor used to trace _and_ what appears before
linear

Differential Revision: [D45713915](https://our.internmc.facebook.com/intern/diff/D45713915/)

kimishpatel committed:
[PT2][Quant] Move to module partitioner for linear pattern quantization

Subgraph matcher is somewhat unreliable as the pattern can vary depending on
the dimensionality of input tensor used to trace _and_ what appears before
linear

Differential Revision: [D45713915](https://our.internmc.facebook.com/intern/diff/D45713915/)

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -458,74 +460,54 @@ def _annotate_conv2d(
+                        raise ValueError(
+                            "Could not find an user of act node within matched pattern."
+                        )
+                    if _is_annotated([act_use_node]) is False:  # type: ignore[list-item]
```
nit: `not _is_annotated([act_use_node])`

jerryzh168 commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -458,74 +460,54 @@ def _annotate_conv2d(
+                            get_act_obs_or_fq_ctr(quantization_config),
+                        )
+                    if bias_node and _is_annotated([bias_node]) is False:
+                        _annotate_output_qspec(
```
are we still annotating output? or do you plan to change it to input_qspec_map?

kimishpatel replied:
are you talking about bias quantization using derived quantization? Otherwise it doesnt really matter? For weight quantization this PR still uses output annotation

kimishpatel committed:
Update on "[PT2][Quant] Move to module partitioner for linear pattern quantization"


Subgraph matcher is somewhat unreliable as the pattern can vary depending on
the dimensionality of input tensor used to trace _and_ what appears before
linear

Differential Revision: [D45713915](https://our.internmc.facebook.com/intern/diff/D45713915/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2][Quant] Move to module partitioner for linear pattern quantization"


Subgraph matcher is somewhat unreliable as the pattern can vary depending on
the dimensionality of input tensor used to trace _and_ what appears before
linear

Differential Revision: [D45713915](https://our.internmc.facebook.com/intern/diff/D45713915/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2][Quant] Move to module partitioner for linear pattern quantization"


Subgraph matcher is somewhat unreliable as the pattern can vary depending on
the dimensionality of input tensor used to trace _and_ what appears before
linear

Differential Revision: [D45713915](https://our.internmc.facebook.com/intern/diff/D45713915/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2][Quant] Move to module partitioner for linear pattern quantization"


Subgraph matcher is somewhat unreliable as the pattern can vary depending on
the dimensionality of input tensor used to trace _and_ what appears before
linear

Differential Revision: [D45713915](https://our.internmc.facebook.com/intern/diff/D45713915/)

[ghstack-poisoned]
